odilo.py

```diff
@@ -0,0 +1,581 @@
+# coding=utf-8
+import base64
+import datetime
+import isbnlib
+import json
+import logging
+
+from sqlalchemy.orm.session import Session
+
+from httplib import HTTPException
+
+from model import (
+    get_one_or_create,
+    Collection,
+    Contributor,
+    Credential,
+    DataSource,
+    DeliveryMechanism,
+    Edition,
+    ExternalIntegration,
+    Hyperlink,
+    Identifier,
+    Representation,
+    Subject,
+)
+
+from core.util.http import RequestNetworkException
+
+from core.analytics import Analytics
+
+from metadata_layer import (
+    CirculationData,
+    ContributorData,
+    FormatData,
+    IdentifierData,
+    Metadata,
+    LinkData,
+    ReplacementPolicy,
+    SubjectData,
+)
+
+from coverage import (
+    BibliographicCoverageProvider,
+)
+
+from config import (
+    CannotLoadConfiguration,
+)
+
+from testing import DatabaseTest
+
+from util.http import (
+    HTTP,
+    BadResponseException,
+)
+
+
+class OdiloAPI(object):
+    log = logging.getLogger("Odilo API")
+
+    LIBRARY_API_BASE_URL = u"library_api_base_url"
+    BASE_URL = "http://localhost:8080/api/v2"  # Debug value by default
+
+    # --- OAuth ---
+    TOKEN_ENDPOINT = BASE_URL + "/token"
+
+    # --- Discovery API ---
+    ALL_PRODUCTS_ENDPOINT = BASE_URL + "/records"
+
+    RECORD_METADATA_ENDPOINT = BASE_URL + "/records/{recordId}"
+    RECORD_AVAILABILITY_ENDPOINT = BASE_URL + "/records/{recordId}/availability"
+
+    # --- Circulation API ---
+    CHECKOUT_ENDPOINT = BASE_URL + "/records/{recordId}/checkout"
+    CHECKOUT_GET = BASE_URL + "/checkouts/{checkoutId}"
+    CHECKIN_ENDPOINT = BASE_URL + "/checkouts/{checkoutId}/return"
+    # Downloads given checkout offering the url where to consume the digital resource.
+    CHECKOUT_URL_ENDPOINT = BASE_URL + "/checkouts/{checkoutId}/download"
+
+    PLACE_HOLD_ENDPOINT = BASE_URL + "/records/{recordId}/hold"
+    HOLD_GET = BASE_URL + "/holds/{holdId}"
+    RELEASE_HOLD_ENDPOINT = BASE_URL + "/holds/{holdId}/cancel"
+
+    PATRON_CHECKOUTS_ENDPOINT = BASE_URL + "/patrons/{patronId}/checkouts"
+    PATRON_HOLDS_ENDPOINT = BASE_URL + "/patrons/{patronId}/holds"
+
+    def update_endpoints_url(self):
```

Comment:
Rather than redefining all these class constants whenever a new BASE_URL is set, I would define the constants once in the class without reference to BASE_URL, and then prepend the collection's base URL before making the request (this could go into do_get and do_post, for example).

Reply:
ok, thank you, I have already changed it.